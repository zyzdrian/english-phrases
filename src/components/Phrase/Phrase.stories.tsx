import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, text, number, array } from '@storybook/addon-knobs';

import { Phrase } from './Phrase';

// eslint-disable-next-line no-undef
const stories = storiesOf('Components', module);
stories.addDecorator(withKnobs);

const DEFAULT_CHAPTER_ID = 2;
const DEFAULT_BUCKET_INDEX = 0;

stories.add(
    'Phrase',
    () => {
        return (
            <Phrase
                phrase={{
                    id: text('id', '1'),
                    appropriate: text('appropriate', 'Use in any situation'),
                    name: text('name', '2 B OR NOT 2 B'),
                    meanings: array('meanings', ['To be or not to be']),
                    chapterId: number('chapterId', DEFAULT_CHAPTER_ID),
                    bucketIndex: number('bucketIndex', DEFAULT_BUCKET_INDEX)
                }}
            />
        );
    },
    { info: { inline: true } }
);
