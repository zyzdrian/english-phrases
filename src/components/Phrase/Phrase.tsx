import React from 'react';
import styled from 'styled-components';
import { FlipCard } from '..';
import { IPhrase } from '../../models';

export enum PHRASE_TEST_IDS {
    name = 'name',
    meaning = 'meaning'
}

const StyledPhrase = styled.div`
background-color: ${props => props.theme.cardBackground};
color: ${props => props.theme.fontColorDark};
transition: background-color ease 0.3s;
padding: 36px;
height: 240px;
max-height: calc(100vh - 184px); // 56px + 2x 24px (bottom menu) + 56px + 24px (top menu)
display: flex;
flex-direction: column;
justify-content: center;
border-radius: 16px;
`;

const StyledFront = styled(StyledPhrase)`
  align-items: center;
`;

const StyledName = styled.div`
font-size: 16px;
font-weight: bold;
text-align: center;
`;

const StyledBack = styled(StyledPhrase)`
  align-items: flex-start;
`;

interface IStyledMeaning {
    isMultiple: boolean;
}

const StyledMeaning = styled.div<IStyledMeaning>`
display: ${props => props.isMultiple ? 'list-item' : 'block'};
padding-bottom: 10px;
&:last-child {
  padding-bottom: 0;
}
`;

export const Phrase: React.FC<IPhraseProps> = ({ phrase, isFlipped, onShow }) => {
    const isMultiple = phrase.meanings.length > 1;
    return (
            <FlipCard
                front={(
                    <StyledFront>
                        <StyledName data-testid={PHRASE_TEST_IDS.name}>{phrase.name}</StyledName>
                    </StyledFront>
                )}
                back={(
                    <StyledBack>
                        {phrase.meanings.map((meaning, index) => (
                            <StyledMeaning key={meaning} isMultiple={isMultiple} data-testid={`${PHRASE_TEST_IDS.meaning}-${index}`}>
                                {meaning}
                            </StyledMeaning>
                        ))}
                    </StyledBack>
                )}
                isOutsidedFlipped={isFlipped}
                onShow={onShow}
            />);
};

export interface IPhraseProps {
    phrase: IPhrase;
    isFlipped?: boolean;
    onShow?: () => void;
}
