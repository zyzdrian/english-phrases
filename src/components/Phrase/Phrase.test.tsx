import React from 'react';
import { cleanup, render } from '@testing-library/react';
import { Phrase, IPhraseProps, PHRASE_TEST_IDS } from './Phrase';
import '@testing-library/jest-dom/extend-expect';

const FIRST_MEANING_INDEX = 0;

describe('<Phrase /> tests', () => {
    const INITIAL_PROPS: IPhraseProps = {
        phrase: {
            id: '1',
            name: 'TEST',
            meanings: ['Test meaning'],
            chapterId: 1,
            appropriate: 'Use in any situation',
            bucketIndex: 1
        }
    };
    let props = { ...INITIAL_PROPS };
    beforeEach(() => {
        props = { ...INITIAL_PROPS };
    });
    afterEach(cleanup);

    it('should render as default phrase', () => {
        const { container, getByTestId } = render(<Phrase {...props} />);

        expect(container.firstChild).toMatchSnapshot();
        expect(getByTestId(PHRASE_TEST_IDS.name)).toHaveTextContent(props.phrase.name);
        expect(getByTestId(`${PHRASE_TEST_IDS.meaning}-0`)).toHaveTextContent(props.phrase.meanings[FIRST_MEANING_INDEX]);
        expect(getByTestId(`${PHRASE_TEST_IDS.meaning}-0`)).toHaveStyle('display: block');
    });

    it('should render multiple meanings', () => {
        props.phrase.meanings = ['First meaning', 'Second meaning'];

        const { container, getByTestId } = render(<Phrase {...props} />);

        expect(container.firstChild).toMatchSnapshot();
        props.phrase.meanings.forEach((meaning, index) => {
            expect(getByTestId(`${PHRASE_TEST_IDS.meaning}-${index}`)).toHaveTextContent(meaning);
            expect(getByTestId(`${PHRASE_TEST_IDS.meaning}-${index}`)).toHaveStyle('display: list-item');
        });
    });
});
