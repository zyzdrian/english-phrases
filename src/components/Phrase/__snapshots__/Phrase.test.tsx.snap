// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Phrase /> tests should render as default phrase 1`] = `
<div
  class="sc-bxivhb eRFxNn"
>
  <div
    class="sc-ifAKCX jJulbc"
  >
    <div
      class="sc-EHOje kGzamQ"
      data-testid="front"
    >
      <div
        class="sc-VigVT sc-jTzLTM YRJrv"
      >
        <div
          class="sc-fjdhpX jxWowm"
          data-testid="name"
        >
          TEST
        </div>
      </div>
    </div>
    <div
      class="sc-bZQynM eprqWe"
      data-testid="back"
    >
      <div
        class="sc-VigVT sc-jzJRlG kpvmUn"
      >
        <div
          class="sc-cSHVUG dhAxXz"
          data-testid="meaning-0"
        >
          Test meaning
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Phrase /> tests should render multiple meanings 1`] = `
<div
  class="sc-bxivhb eRFxNn"
>
  <div
    class="sc-ifAKCX jJulbc"
  >
    <div
      class="sc-EHOje kGzamQ"
      data-testid="front"
    >
      <div
        class="sc-VigVT sc-jTzLTM YRJrv"
      >
        <div
          class="sc-fjdhpX jxWowm"
          data-testid="name"
        >
          TEST
        </div>
      </div>
    </div>
    <div
      class="sc-bZQynM eprqWe"
      data-testid="back"
    >
      <div
        class="sc-VigVT sc-jzJRlG kpvmUn"
      >
        <div
          class="sc-cSHVUG iwqkMh"
          data-testid="meaning-0"
        >
          First meaning
        </div>
        <div
          class="sc-cSHVUG iwqkMh"
          data-testid="meaning-1"
        >
          Second meaning
        </div>
      </div>
    </div>
  </div>
</div>
`;
